### ISSUE TITLE
Strange output of Flan-T5-Large when using float16 converted model

### ISSUE THREAD

**Issue by Matthieu-Tinycoaching** [OpenNMT/CTranslate2#1074]
Hi,

I've installed the "python-wheels" artifact from #1066 .

**Comment by guillaumekln** [comment]
Thanks again for reporting the issue.

I found that the model produces NaN values after the 5th layer. Apparently this is a known issue when running T5 models in FP16: https://github.com/huggingface/transformers/pull/10956

We can try the change that is proposed in this PR but it does not seem efficient. Maybe you can go one step further and use the `int8` quantization for this model?

**Comment by Matthieu-Tinycoaching** [comment]
Ok right 👍

**Comment by baptistejamin** [comment]
Update: After investigating, it seems a part of the answer is there: https://github.com/huggingface/transformers/pull/22095/commits

**Comment by redthing1** [comment]
Yeah, I get it returning its own input many times. No real answers.

**Comment by redthing1** [comment]
Here are more details: https://github.com/huggingface/transformers/issues/20287#issuecomment-1342219429

**Comment by redthing1** [comment]
@guillaumekln This appears to be the necessary patch: https://github.com/larsmennen/transformers/commit/f90b269505a02d5476661ad77cb426e44e7561d3

**Comment by guillaumekln** [comment]
#1239 keeps the FFN output layer in FP32 as suggested in the different issues mentioned above.

Can you help testing this development?

1. Go to this [build page](https://github.com/OpenNMT/CTranslate2/actions/runs/5057553803)
1. Download the artifact "python-wheels"
1. Extract the archive
1. Install the wheel matching your system and Python version with `pip install --force-reinstall`

The model must converted again with this version.

**Comment by redthing1** [comment]
@guillaumekln Thank you for the quick fix! This appears to solve the problem for me.

**Comment by baptistejamin** [comment]
We are running the tests as we speak on a FlanT5 XXL.

I seems there is the same problem in fp8. I will try with fp16. Maybe related to this ? https://github.com/OpenNMT/CTranslate2/pull/1239/files#diff-c2632813f7cc9ff44bda20479812ebefd15be7f7fe1bd099e553111bcb6750acR171

**Comment by guillaumekln** [comment]
> I seems there is the same problem in fp8.

You meant int8, right? More specifically did you try int8 or int8_float16?

**Comment by baptistejamin** [comment]
Sorry about the typo. Int8 indeed.

So here at the tests with Flan T5 XXL:

`pip install ctranslate2-3.13.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl`

With `ct2-transformers-converter --model google/flan-t5-xxl --output_dir flan-xxl-c2`

`Who is Barrack Obama?:`

int8: `us president/barackhopatsumibi/podium/barhamambua00barackoblamesurprize/summonant/218974312`
int8_float6: Error: `ValueError: expected storage to be of type float32, but is of type float16`
float16: Error: `ValueError: expected storage to be of type float32, but is of type float16`

With `ct2-transformers-converter --model google/flan-t5-xxl --output_dir flan-xxl-c2 --quantization float16`

int8: `president united states pubic servant elect`
int8_float6: Error: `ValueError: expected storage to be of type float32, but is of type float16`
float16: Error: `ValueError: expected storage to be of type float32, but is of type float16`

I may be missing something.

**Comment by bradfox2** [comment]
Running a domain specific t5-3b and trying @baptistejamin  prompt with latest build: https://github.com/OpenNMT/CTranslate2/actions/runs/5068556228 :

Who is barrack obama? -> Barack Obama is the 44th President of the United States. He was elected in 2008 and has been in office since January 20, 2009.

With: 
translate_args = {
    "repetition_penalty": 1.3,
    "sampling_topk": 1,
    "beam_size": 1,
    "sampling_temperature": 0.3,
    "max_decoding_length": 512,
}

ctranslate2.Translator(
    device="cpu",
    inter_threads=8,
    intra_threads=8,
    compute_type="int8",
)

Equally response on: 
    device="cuda",
    inter_threads=8,
    intra_threads=8,
    compute_type="int8_float16",

**Comment by redthing1** [comment]
Fixed for me as well, thank you

**Comment by baptistejamin** [comment]
I can confirm it works fine with Flan XL int8 with the latest build

However, Flan XXL int8 it still returns non-sensical text, and it for the same prompt, the output is worse than a XL model.

Finally, float16 mode still returns `<pad><pad><pad><pad>...` for both Flan XL and Flan XXL official models.

**Comment by Matthieu-Tinycoaching** [comment]
Hi,
Is this corrected build published on the last release ?

**Comment by bradfox2** [comment]
I've been working with XXL in int8 with 3.15.1 and it appears to be emitting valid responses.  Both cuda and CPU.  @baptistejamin FYI

**Comment by baptistejamin** [comment]
What’s your prompt and parameters ?Sent from my iPhoneOn 15 Jun 2023, at 03:55, Bradley Fox ***@***.***> wrote:﻿
I've been working with XXL in int8 with 3.15.1 and it appears to be emitting valid responses.  Both cuda and CPU.  @baptistejamin FYI

**Comment by bradfox2** [comment]
@baptistejamin  Any prompt works just fine - However, I had a recent converted model fail in the same way.  It that case it was how tokenizer was converted to the config.json that that CT2 creates.   The bos and and decoder start sequence tokens were </s>, which is wrong for T5.  

I haven't spent much time troubleshooting but it comes from when the Seq2Seq model inits in model_spec.py: 386

```
class SequenceToSequenceModelConfig(ModelConfig):
    """Configuration for sequence-to-sequence models."""

    def __init__(
        self,
        unk_token: str = "<unk>",
        bos_token: str = "<s>",
        eos_token: str = "</s>",
        decoder_start_token: Optional[str] = "<s>",
        add_source_bos: bool = False,
        add_source_eos: bool = False,
    ):
```

Maybe this is also your issue?

**Comment by guillaumekln** [comment]
FYI, CTranslate2 3.17 supports the "bfloat16" compute type which should be used instead of "float16" for these models.

**Comment by tsuyoshi-fukuzawa** [comment]
When I use T5 fine-tuned with bf16 and convert it to float16 with CTranslate, the inference results are all PAD tokens.
When using int8_float16, it returns the correct results. However, with bfloat16, all hypotheses have the same value.
If the T5 model is fine-tuned without bf16, float16 works properly. Therefore, I decided to use T5 fine-tuned without bf16.
